crates/wasmtime/src/compile/scc.rs

```diff
@@ -0,0 +1,528 @@
+    /// members of it.
+    ///
+    /// Iteration happens in reverse-topological order (successors are visited
+    /// before predecessors in the resulting SCC DAG).
+    pub fn iter(&self) -> impl ExactSizeIterator<Item = (Component, &[Node])> + '_ {
+        self.components.iter().map(|(component, range)| {
+            let start = usize::try_from(range.start).unwrap();
+            let end = usize::try_from(range.end).unwrap();
+            (component, &self.component_nodes[start..end])
+        })
+    }
+
+    /// Iterate over each strongly-connected component.
+    ///
+    /// Iteration happens in reverse-topological order (successors are visited
+    /// before predecessors in the resulting SCC DAG).
+    pub fn keys(&self) -> impl ExactSizeIterator<Item = Component> {
+        self.components.keys()
+    }
+
+    /// Iterate over the `Node`s that make up each strongly-connected component.
+    ///
+    /// Iteration happens in reverse-topological order (successors are visited
+    /// before predecessors in the resulting SCC DAG).
+    pub fn values(&self) -> impl ExactSizeIterator<Item = &[Node]> + '_ {
+        self.components.values().map(|range| {
+            let start = usize::try_from(range.start).unwrap();
+            let end = usize::try_from(range.end).unwrap();
+            &self.component_nodes[start..end]
+        })
+    }
+
+    /// Get the `Node`s that make up the given strongly-connected component.
+    pub fn nodes(&self, component: Component) -> &[Node] {
+        let range = self.components[component].clone();
+        let start = usize::try_from(range.start).unwrap();
+        let end = usize::try_from(range.end).unwrap();
+        &self.component_nodes[start..end]
+    }
+}
+
+/// An iterative depth-first traversal.
+struct Dfs<Node> {
+    stack: Vec<DfsEvent<Node>>,
+}
+
+impl<Node> Dfs<Node> {
+    fn new(roots: impl IntoIterator<Item = Node>) -> Self {
+        Self {
+            stack: roots.into_iter().map(|v| DfsEvent::Pre(v)).collect(),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+enum DfsEvent<Node> {
+    /// The first time seeing this node.
+    Pre(Node),
+
+    /// After having just visited the given edge.
+    AfterEdge(Node, Node),
+
+    /// Finished visiting this node and all of its successors.
+    Post(Node),
+}
+
+impl<Node> Dfs<Node>
+where
+    Node: Copy + std::fmt::Debug,
+{
+    /// Pump the traversal, yielding the next `DfsEvent`.
+    fn next<S>(
+        &mut self,
+        successors: impl Fn(Node) -> S,
+        seen: impl Fn(Node) -> bool,
+    ) -> Option<DfsEvent<Node>>
+    where
+        S: Iterator<Item = Node>,
+    {
+        loop {
+            let event = self.stack.pop()?;
+
+            if let DfsEvent::Pre(node) = event {
+                if seen(node) {
+                    continue;
+                }
+
+                self.stack.push(DfsEvent::Post(node));
+                self.stack.extend(successors(node).flat_map(|v| {
```

Comment:
This is a neat iterator chain but I wonder if it would be clearer in imperative style in this case (vs. the chaining and nested if-expression returning an option) -- something like:

```rust
for successor in successors(node) {
  self.stack.push(DfsEvent::AfterEdge(node, successor));
  if !seen(successor) {
    self.stack.push(DfsEvent::Pre(successor));
  }
}
```

Reply:
Yeah this is a reflection of how it was originally written and then modified as I found and fixed bugs. Should have moved it over to pushes already.